Client/Frontend/Browser/Playlist/Cache/PlaylistDownloadManager.swift

```diff
@@ -0,0 +1,485 @@
+// Copyright 2020 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import Foundation
+import AVFoundation
+import Shared
+
+private let log = Logger.browserLogger
+
+protocol PlaylistDownloadManagerDelegate: class {
+    func onDownloadProgressUpdate(id: String, percentComplete: Double)
+    func onDownloadStateChanged(id: String, state: PlaylistDownloadManager.DownloadState, displayName: String)
+}
+
+private protocol PlaylistStreamDownloadManagerDelegate: class {
+    func localAsset(for pageSrc: String) -> AVURLAsset?
+    func displayNames(for mediaSelection: AVMediaSelection) -> String
+    func onDownloadProgressUpdate(id: String, percentComplete: Double)
+    func onDownloadStateChanged(id: String, state: PlaylistDownloadManager.DownloadState, displayName: String)
+}
+
+struct MediaDownloadTask {
+    let id: String
+    let name: String
+    let asset: AVURLAsset
+    
+    enum Keys: String {
+        case id
+        case state
+        case displayName
+    }
+}
+
+public class PlaylistDownloadManager: PlaylistStreamDownloadManagerDelegate {
+    private let hlsSession: AVAssetDownloadURLSession
+    private let fileSession: URLSession
+    private let hlsDelegate: PlaylistHLSDownloadManager
+    private let fileDelegate: PlaylistFileDownloadManager
+    private var didRestoreSession = false
+    
+    weak var delegate: PlaylistDownloadManagerDelegate?
+    
+    public enum DownloadState: String {
+        case downloaded
+        case inProgress
+        case invalid
+    }
+    
+    init() {
+        hlsDelegate = PlaylistHLSDownloadManager()
+        fileDelegate = PlaylistFileDownloadManager()
+        
+        let hlsConfiguration = URLSessionConfiguration.background(withIdentifier: "com.brave.playlist.hls.background.session")
+        hlsSession = AVAssetDownloadURLSession(configuration: hlsConfiguration,
+                                               assetDownloadDelegate: hlsDelegate,
+                                               delegateQueue: OperationQueue())
+        
+        let fileConfiguration = URLSessionConfiguration.background(withIdentifier: "com.brave.playlist.file.background.session")
+        fileSession = URLSession(configuration: fileConfiguration,
+                                 delegate: fileDelegate,
+                                 delegateQueue: OperationQueue())
+        
+        hlsDelegate.delegate = self
+        fileDelegate.delegate = self
+    }
+    
+    func restoreSession(_ completion: @escaping () -> Void) {
+        // Called from AppDelegate to restore pending downloads
+        guard !didRestoreSession else { return }
+        didRestoreSession = true
+        
+        let group = DispatchGroup()
```

Comment:
Do you think it might a better idea to use semaphor here instead of group so we can limit number of downloads that can restore at once( just to be on the safe side)

Reply:
No because this is used for multiple requests. It is used for `hlsDelegate.restoreSession` and `fileDelegate.restoreSession` which can both run in parallel. Using the `semaphore` will block the current thread and force them to run synchronously.